lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -316,10 +316,44 @@ llvm::Error HostManager::runNetworkBlocking(llvm::StringRef networkName,
   return std::move(*DCHECK_NOTNULL(runErr.get()));
 }
 
+void HostManager::dispatchNextRun() {
+
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  if (inferQueue_.size()) {
+    // Get the next request, unfortunately priority_queue only
+    // provides a const ref to the top element, since we need to move
+    // it we first cast it to remove the const.
+    auto request = std::move(const_cast<InferRequest &>(inferQueue_.top()));
+    inferQueue_.pop();
+    executor_->run(
+        networks_[request.networkName].dag.root.get(),
+        std::move(request.context), request.requestID,
+        [this, callback = request.callback, name = request.networkName](
+            RunIdentifierTy runID, llvm::Error err,
+            std::unique_ptr<ExecutionContext> context) {
+          {
+            std::lock_guard<std::mutex> networkLock(networkLock_);
+            auto it = networks_.find(name);
+            if (it != networks_.end()) {
+              it->second.refcount--;
+            }
+          }
+          TRACE_EVENT_INSTANT(context->getTraceContext(), TraceLevel::RUNTIME,
+                              "finish_" + name);
+          callback(runID, std::move(err), std::move(context));
+          dispatchNextRun();
```

Comment:
The fact that we `dispatchNextRun` from inside a callback created by `dispatchNextRun` bugs me.  Is this, like, a common pattern in concurrent code?

It feels like what we really want is for the HostManager to push tasks into the priority queue, and to have a priority-aware ThreadPoolExecutor that knows how to pull the top pri item in its event loop.

cc @tracelogfb, just in case I'm out of my depth in concurrent program design...